feat: track content origins independent of files

kopecs opened the pull request:
Currently we use a combination of `Fpath.t` and `string` pervasively for a couple distinct purposes:
- reporting where a finding originated and
- getting the contents of the object we’re scanning (e.g., with UFile.read_file).

This dual-purpose use means it is difficult to:
- create a target where the contents are not the verbatim file contents
- associate additional data to findings about location which we might want to report to users(e.g., “nested within a metavariable pattern”; “via extract mode”)
- scan anything which is not resident on the filesystem (e.g., metavariable pattern, git object)---currently this *requires* writing a tempfile.

Mainly, this PR adds one additional type, ~`Source.t`~ `Origin.t`, which acts as a "user facing" location we can report. In order to utilize this, we also add several additional types (See ~`Target_location`~ `Target`) which serve to replace `Input_to_core.t` and similar inside of the engine. Conversion from prior types to these new types is added where needed. The types in ~`Target_location`~ `Target` then have a field ~`source : Source.t`~ `origin : Origin.t` to attach the additional information.

In this PR, ~`Source.t`~ `Origin.t` only has a single variant, but I would anticipate adding more in the future (for instance, `GitBlob` for historical scanning).

Additionally, a file path is currently present in these types for the existing pattern of writing data to a tempfile for the purpose of scanning. However, I think we can eventually get rid of the file path for instances where there is not an advantage to writing the content to disk, but I don't do so here.

See also discussion in https://semgrepinc.slack.com/archives/C05STGMM53Q/p1706034545394119

---

For reviewers: I think this is best reviewed commit-by-commit; I have tried to put up a rather clean history.

kopecs committed:
Create libs/source

Defines a `Source.t` type; currently just for files. This can be used to
replace Fpath.t or similar types (e.g., string) where values of that
type designate a target's location in a way which the user can
understand (i.e., not a tempfile).

kopecs committed:
Add `source` field to `Xtarget.t`

This allows us to record a user-facing source for where the target came
from. Currently this is just a file, but could be expanded in the
future. Right now, it enables cases where we might have a tempfile but
want to report something at a different file for the user. In the
future, this could be used to report, e.g., a git object, network
request, more granular treatment of nested content or something else as
the source for content.

kopecs committed:
Add List_.sort_by_key

Allows for sorting a list by a mapped key. The key is computed only once
for each value in the list and is not retained after sorting.

kopecs committed:
Move Core_scan.xtarget_of_file to Xtarget.resolve

kopecs committed:
Use Target_location.t in target_source

Formerly we used In.target here, but this is more restrictive than we
would like in tracking source info and is tied to the external json
input format. This requires some additional translation, but we were
already shuffling data around in most instances, so this just changes
lambdas for some maps.

kopecs committed:
Use source for xpattern{regex,alien,space} matches

kopecs committed:
Add Target_location module

This is intended to be an expanded version of the Input_to_core.target
type which can also track source information and is more easily
convertible into an Xtarget (potentially "Target_content" or similar in
the future). Martin has suggested:

    type target_contents = {
      (* to be used in error messages: file path, URL, stdin, ... *)
      source: Source.t;
      (* path to a regular file which may be temporary *)
      path: Fpath.t;
      (* file contents *)
      contents: string;
      (* + contents split into lines or whatever is convenient *)
    }

which aligns closely with our current Xtarget.t.

kopecs committed:
Use `Target_location.t` in Core_scan

This replaces earlier usage of `In.target`. Places where we dealt with
`In.target`s are updated to convert to `Target_location.t`s. This preps
for being able to thread source information through and have better
reporting to users of locations.

kopecs committed:
Add source information to Pattern_match.t

Now we can actually leverage the source information we are storing in,
e.g., `Target_location.t`. This updates `Pattern_match.t` to have an
associated source and to use it in generating the json output. Some
changes to the Xpattern_match code is required: mostly, a source is
carried along with the file itself. We could consider just using
`Target_location.t` here, but it was unclear to me how much effort is
required.

kopecs commented:
Pytest failure seems like I flipped the sort order by accident (makes sense).

kopecs committed:
fixup! Use `Target_location.t` in Core_scan

Flips sort order so that highest is first, as we want.

kopecs committed:
fixup! Use Target_location.t in target_source

mmcqd commented on dune-project:
```diff
@@ -368,6 +368,10 @@ functions.
   )
 )
+(package (name source)
+  (synopsis "XXX")
```
Might as well actually write something here no?

kopecs replied:
Done in af778e761cbbc7218c2a55ec906a7ed1b3aebd93

mjambon replied:
I still see XXX.

kopecs replied:
I was just waiting to finish addressing Matthew's comments to push the changes. It should be present now.

mmcqd commented on src/core_scan/Core_scan.ml:
```diff
@@ -921,8 +954,9 @@ let select_applicable_rules_for_target ~analyzer ~products ~path
          match r.R.paths with
-         | Some paths when respect_rule_paths ->
-             Filter_target.filter_paths paths path
+         | Some paths when respect_rule_paths -> (
```
Should we be doing this based on the source? What if the source is some relative path that doesn't get correctly filtered when an absolute path would?

kopecs replied:
Yes, this should be based on the source so that when we have non-file sources we can correctly apply include/exclude to them rather than the tempfile path. We already implicitly add a `**/` and `/**` before and after (resp.) the pattern, and this is documented to be relative to the project root (https://semgrep.dev/docs/writing-rules/rule-syntax/#paths), so I believe this is sound.

As a motivating example for this, consider a git-based source:
```ocaml
| GitBlob { paths = target_paths; _ } ->
    List.exists
      (fun (_commit_sha, path_at_commit) ->
        Filter_target.filter_paths paths path_at_commit)
      target_paths)
```
where we want to run if the blob is present at some commit with a path indicating it should be run against. If we ran against the file field then we wouldn't be able to have this behaviour.

kopecs replied:
Added a comment in 72603150d9f141f50d0ac51a07fef6c9bd2bc2ce

mmcqd commented on src/core_scan/Core_scan.ml:
```diff
@@ -940,12 +974,10 @@ let mk_target_handler (config : Core_scan_config.t) (valid_rules : Rule.t list)
+      let lockfile_target = lockfile_target_of_location lockfile_location in
       let applicable_supply_chain_rules =
-        select_applicable_supply_chain_rules ~lockfile_kind ~path:lockfile
+        select_applicable_supply_chain_rules ~lockfile_kind:kind ~path:file
```
Here we're filtering by path but in the one for code targets we're filtering by source

kopecs replied:
Ah, good catch. I'll update the supply chain stuff too.

kopecs replied:
Fixed in 72603150d9f141f50d0ac51a07fef6c9bd2bc2ce

mmcqd commented on src/matching/Match_patterns.ml:
```diff
@@ -139,7 +139,8 @@ let (rule_id_of_mini_rule : Mini_rule.t -> Pattern_match.rule_id) =
   }
 
-let match_rules_and_recurse m_env (file, hook, matches) rules matcher k any x =
+let match_rules_and_recurse m_env (file, source, hook, matches) rules matcher k
```
A 4-tuple is getting a bit ugly, should we make this a record?

kopecs replied:
I agree, this is not super clean. I was looking for a smallish diff here, but just making this a record wouldn't be too bad.

mjambon replied:
I don't mind if the diff gets big after the tricky parts have been resolved.

kopecs replied:
Done in bdb51b197b7ee4fc901ad1f3d34313f2e25b10f0

mmcqd commented on src/reporting/Core_json_output.ml:
```diff
@@ -223,23 +223,23 @@ let unsafe_match_to_match
+      | File path ->
+          if
+            (!!path <> min_loc.pos.file || !!path <> max_loc.pos.file)
+            && min_loc.pos.file <> "FAKE TOKEN LOCATION"
```
Isn't there a `Tok.is_fake` function or something like this?

kopecs replied:
This was the existing code (this is just a move). I can see if there's an improvement but I am not confident about test coverage for this as is (cc: @brandonspark ?) so I hesitate to change it here.

brandonspark replied:
If there is a `Tok.is_fake` function, I may have just been unaware of it when I originally wrote this code.

mmcqd commented on src/reporting/Core_json_output.ml:
```diff
@@ -223,23 +223,23 @@ let unsafe_match_to_match
-    path = Fpath.v file;
+    path =
+      (match x.source with
+      (* We need to do this, because in Terraform, we may end up with a `file` which
```
Would be useful to explain why and in what way we create a fake/erroneous file

kopecs replied:
Yeah, as above this is just moved. I don't have any additional context beyond this comment cc: @brandonspark who wrote this I think.

brandonspark replied:
This is because in Terraform, all the files within a particular directory are actually just one big file, semantically. We process it as such, putting all the original file's contents into a fake file, but this causes fake tokens to appear.

The comment could be better.

mmcqd commented on src/core_scan/Core_scan.ml:
```diff
@@ -179,14 +179,14 @@ let sort_code_targets_by_decreasing_size (targets : Target_location.code list) :
   |> List_.sort_by_key
        (fun (target : Target_location.code) -> UFile.filesize target.file)
-       Int.compare
+       (Fun.flip Int.compare)
```
lol. We should add a comment like "We want these in *descending* order"

kopecs replied:
Sure.

kopecs replied:
Done in c6fb06e5107f09a5caca7a57a18e5421aacb9f58

mmcqd reviewed (approved):
big moves

kopecs committed:
Improve package metadata for source

mjambon commented on src/core/Target_location.ml:
```diff
@@ -0,0 +1,82 @@
+   *)
+  lockfile : lockfile option;
+}
+[@@deriving show]
```
If the definitions aren't mutually recursive, write them sequentially without using `and` (otherwise I spend 10 s trying to figure out why and where they're mutually recursive only to discover that they're not).

kopecs replied:
I can change it here. Do you have the same objection to this in the interface? I prefer having the "main topic" be first, as we do in, e.g., `Pattern_match.mli` both for reading and for the generated documentation.

kopecs replied:
Done for the ml file in 53d1290887b2dfd54da804bf66665f5f3336c3d1

mjambon commented on src/core/Target_location.ml:
```diff
@@ -0,0 +1,82 @@
+(* Cooper Pierce
```
You need to explain what this code is about. Always.

kopecs replied:
Done in 53d1290887b2dfd54da804bf66665f5f3336c3d1

mjambon commented on src/core/Target_location.mli:
```diff
@@ -0,0 +1,72 @@
+(* Cooper Pierce
```
Same remark as above, a reader needs to know quickly what this module is about from reading the first few lines of the file.

emjin replied:
I think it's pretty common for our mli files to not have file header summary comments. I agree that the ml files need preludes.

kopecs replied:
Done in 53d1290887b2dfd54da804bf66665f5f3336c3d1

mjambon commented on libs/source/Source.ml:
```diff
@@ -0,0 +1,28 @@
+type t = File of Fpath.t [@@deriving show, eq, ord]
```
Did you intend to add other kinds of sources? You need to describe what this module is about and add an mli.

kopecs replied:
mli added in 3ef8b07eb

kopecs replied:
Docs in 7347e9833509121bdd69d72c7f88dd7cdc922577

mjambon commented on src/core/Target_location.ml:
```diff
@@ -0,0 +1,82 @@
+  kind : Lockfile_kind.t;
+  manifest : manifest option;
+}
+[@@deriving show]
```
This probably should have its own module.

kopecs replied:
This seems like a duplicate of/substantially similar to the following comment, so resolving this thread in favour of discussing there.

mjambon commented on src/core/Target_location.ml:
```diff
@@ -0,0 +1,82 @@
+[@@deriving show]
+
+and manifest = { source : Source.t; file : Fpath.t; kind : Manifest_kind.t }
+[@@deriving show]
```
You have to explain what lockfiles and manifests are. This is better done in a dedicated module.

mmcqd replied:
These are explained in `Lockfile_target.ml`, perhaps the comments should be moved or duplicated.

emjin replied:
I would prefer that the types be simply defined in Lockfile_target.ml or that the fields be unified.

kopecs replied:
> I would prefer [...] that the fields be unified.

I did consider defining this and `Lockfile_target` as

```ocaml
type 'content t = {
  source : Source.t;
  file : Fpath.t;
  kind : Lockfile_kind.t;
  manifest : manifest_target option;
  content: 'content;
}
type target_content = {
  lazy_content : string lazy_t;
  lazy_ast_and_errors : Dependency.t list lazy_t;
}
type location = unit t;
type target = target_content t;
```
(and then similar for `code` in `Xtarget` or a substantially similar re-named module)

but this is at odds with having a unified `Target_location.t` type which shares fields, rather than being a sum of `Xtarget.location` and `Lockfile_target.location` as defined above.

Given that, which would you prefer?

emjin replied:
I still prefer the unified `Target_location.t`. My goal is for people changing fields for something specific to their target type (e.g. lockfile) to not think about the generic `Target_location`, and for people changing something generic about the `Target_location` to not think about the possible targets types.

kopecs replied:
Re fields: Addressed in fc35371b4020d3fdf1beb37d3446dbe5f1e81f70. Re explanation: some commit added docs; see full diff.

mjambon commented on src/core/Target_location.ml:
```diff
@@ -0,0 +1,82 @@
+ * LICENSE for more details.
+ *)
+
+type t = Code of code | Lockfile of lockfile [@@deriving show]
```
I wonder why these two things have to be grouped together. What's a target location? What do these files have in common? If they have something in common, then `t` should be a record. If they have nothing in common, what's the point of grouping them under this type?

kopecs replied:
Historically this would have only been one type, currently the `code` type (as `t` instead). However, with @mmcqd's recent changes in https://github.com/semgrep/semgrep/pull/9695, `Input_to_core.target` is now a sum in the ATD. So this is defined similarly by analogy (and to reduce the complexity of when I had rebased my changes on that PR to resolve conflicts).

I can change it so that `Target_location` unifies shared fields, but I was a bit hesitant to do so since I don't know what Matthew is relying on for his changes to pull supply chain into core.

mmcqd replied:
Unifying the shared fields seems like a good idea to me. The description of what these things are is in `Input_to_core.atd`. It definitely kinda feels like we're ending up with too many different types that are all very similar in name and structure, but with slightly different amounts of information.

kopecs replied:
> It definitely kinda feels like we're ending up with too many different types that are all very similar in name and structure, but with slightly different amounts of information.

Yes, I think ideally we would just have one set of these, instead of two, split across here and `Input_to_core`, but while we still need to allow pysemgrep to send the existing list of targets it seems difficult to ditch `Input_to_core` in favour of non-ATD types.

emjin replied:
I am strongly in favor of unifying the shared fields. I'd then like us to have separate modules for the fields for the other products so that `Target_location` can be generic.

kopecs replied:
Wrt unifying the shared fields, I do think it makes sense, but I don't see a straightforward way to do so, since we rely on being able to deal with only `Target_location.code` values in some circumstances, and having this type be a record rather than a sum would remove that capability (at least without a lot of type mangaling at those use sites).

> I'd then like us to have separate modules for the fields for the other products so that Target_location can be generic.

It does sound like there is potentially some confusion that this is related to any particular product. That isn't the intention. Indeed, a supply chain or secrets-only scan would still have "code" targets. This is because the "code" part is not a description of the product, but rather of the sort of thing the target _contains_. Code targets are "code" since they comprise source code; not because they are used with the code/sast product.

emjin replied:
Since I'm sure I'm mentioning this for the first time, I hate the name of the Code product

emjin replied:
> we rely on being able to deal with only Target_location.code values in some circumstances

That's fair. Is it reasonable to do something like this?
```
type 'a target = {
   source: Source.t
   internal_path_to_content: Fpath.t
   details: 'a
}
```

kopecs replied:
Addressed in fc35371b4020d3fdf1beb37d3446dbe5f1e81f70; there still exists a type which performs this grouping but otherwise record fields are shared as much as I could reasonably figure.

mjambon commented on src/core/Lockfile_target.ml:
```diff
@@ -4,11 +4,12 @@
-  lazy_lockfile_ast_and_errors : Dependency.t list lazy_t;
-  manifest_target : manifest_target option;
+  source : Source.t;
+  file : Fpath.t;
```
If `source` is a file, what is `file`? What's its relationship with `source`?

kopecs replied:
I think I have some comments which would answer this in `Target_location` which are more broadly applicable. I'll add something similar here.

kopecs replied:
Done in fdc6557f160487d813c5a16a2177ed02f5f5915e

kopecs committed:
Use source to filter paths for both sca and sast

Also improves source-level comments for documentation

mjambon commented on src/core_cli/Core_command.ml:
```diff
@@ -210,7 +211,7 @@ let semgrep_core_with_one_pattern (config : Core_scan_config.t) : unit =
                        Core_scan.parse_equivalences config.equivalences_file )
                      minirule
-                     (Fpath.v file, lang, ast)
+                     (Fpath.v file, File (Fpath.v file), lang, ast)
```
Call `Fpath.v` as little as possible. It's an expensive operation, unlike `Fpath.to_string`.

kopecs replied:
Done in ed375e0a88c5792bb6c349ccaaf056254ad846ce; see also https://github.com/semgrep/semgrep/pull/9715/files/2193eb567fd8c8fc924cb989e1726c2602275a44#r1481932234 as is relevant more generally.

mjambon commented on src/core_cli/Core_command.ml:
```diff
@@ -188,8 +188,9 @@ let semgrep_core_with_one_pattern (config : Core_scan_config.t) : unit =
-             | `LockfileTarget (t : Input_to_core_t.lockfile_target) -> t.path)
+               | Target_location.Code { file; _ }
+               | Lockfile { file; _ }
+               -> Fpath.to_string file)
```
1. This use case tells me that `Target_location.t` should be a record with a `file` field.
2. Please prefer explicit type annotations to disambiguate constructors because it's more precise and more readable: `(fun (tl : Target_location.t) -> tl.file)`.
3. Conversions from fpath to strings should only occur right before using them e.g. in a printf or when passing them to an external library function that only takes strings - otherwise they have to be reparsed and revalidated later, like below.

kopecs replied:
2. is done in 135dff8da0424f9a79b39d7a72e2c7c10936e403

kopecs replied:
Re: 1. See my comment [here](https://github.com/semgrep/semgrep/pull/9715/files/2193eb567fd8c8fc924cb989e1726c2602275a44#r1480677358). This should use the function `Target_location.file` however; so I have made that update in 4014a2d3f0523b52e9be25b1ceab1123c3b60397.

Edit: Futher updated in fc35371b4020d3fdf1beb37d3446dbe5f1e81f70.

kopecs replied:
3. I do not think this can easily be done here. `file` must be a string for a later call to `Core_error.try_with_print_exn_and_reraise` or `Core_error.try_with_exn_to_error`. Each function calls `Core_error.exn_to_error`. That function contains a check

```ocaml
          let loc =
            (* TODO: we shouldn't build Tok.t w/out a filename, but
               lets do it here so we don't crash until we do *)
            if not String.(equal file "") then Tok.first_loc_of_file file
            else Tok.fake_location
```
which to me implies that `file` may be an empty string by this point. However, if file were instead `Fpath.t`, [we cannot create such a value](https://erratique.ch/software/fpath/doc/Fpath/index.html#val-of_string) generally. At this usage, since the values have already been validated to be valid paths, I think that would be fine, but it would potentially pose issues at other uses of `Core_error.exn_to_error` or functions which call it.

kopecs committed:
Document Fun.flip usage

mjambon commented on src/core_scan/Core_scan.ml:
```diff
@@ -702,57 +706,60 @@ let iter_targets_and_get_matches_and_exn_to_errors (config : Core_scan_config.t)
+let target_location_of_input_to_code (input : In.target) : Target_location.t =
+  match input with
+  | `CodeTarget x -> Code (code_target_location_of_input_to_code x)
+  | `LockfileTarget x -> Lockfile (lockfile_target_location_of_input_to_code x)
```
I don't about this type that uses polymorphic variants but I suspect classic variants are better. I prefer them anyway. The reasons are:
* simplicity by consistency (why use two tools when we could use just one)
* less confusing error messages, especially when type annotations are lacking
Reasons for using polymorphic variants include:
* extensibility, subtyping - which is occasionally useful

kopecs replied:
These are polymorphic variants due to their ATD definitions. @mmcqd introduced these types in https://github.com/semgrep/semgrep/pull/9695; they are not net-new here. If we want to change them to be classic variants instead, I can do so, but I might rather do so in a separate PR since they aren't really related to the changes here.

mmcqd replied:
Yep these are from `Input_to_core.atd`. This should also probably be turned into a record with shared fields and then a field with a sum.

kopecs replied:
I will defer to whatever Matthew wants to do here, but since this would require an additional interfaces change I think it best to do in a separate PR, especially as this PR does not introduce these types. Resolving here.

mjambon commented on src/engine/Match_dependency.ml:
```diff
@@ -51,9 +51,9 @@ let match_dependency_formula :
+ fun { lazy_ast_and_errors; _ } ->
   List.concat_map (fun pat ->
-      match_dependency_pattern (Lazy.force lazy_lockfile_ast_and_errors) pat)
+      match_dependency_pattern (Lazy.force lazy_ast_and_errors) pat)
```
I'd go easy on the lazy computations. Is the laziness really needed? It makes things harder to debug imo.

kopecs replied:
I believe we currently do rely on the AST being lazy in some cases (see the comment in what is now `Xtarget.parse_file`, at least for `Xtarget.t`, not `Lockfile_target.t`). I would defer to @mmcqd who added this in https://github.com/semgrep/semgrep/pull/9695.

mmcqd replied:
I just copied the way that `Xtarget` works, I don't have any thought out need for laziness actually

kopecs replied:
I am resolving this thread here since it is not related to this PR; if changes are desired Matthew can do so in a separate PR.

mjambon commented on src/engine/Test_engine.ml:
```diff
@@ -83,7 +83,8 @@ let xtarget_of_file (xlang : Xlang.t) (target : Fpath.t) : Xtarget.t =
   in
   {
-    Xtarget.file = target;
+    Xtarget.source = File target;
```
You don't need the `Xtarget.` field prefix here. The compiler knows because the return type of the function is set with the annotation `let ... : Xtarget.t = ...`.

kopecs replied:
Sure; I can remove this.

kopecs replied:
Done in 135dff8da0424f9a79b39d7a72e2c7c10936e403

mjambon commented on src/engine/Unit_engine.ml:
```diff
@@ -701,7 +702,8 @@ let tainting_test lang rules_file file =
+               Xtarget.source = File file;
+               file;
                xlang = Xlang.L (lang, []);
                lazy_content = lazy (UFile.read_file file);
```
same here. I would write:
```
let xtarget : Xtarget.t = {
  source = ...; ...
}

kopecs replied:
Done in 135dff8da0424f9a79b39d7a72e2c7c10936e403

mjambon reviewed (dismissed):
Please clarify the relationship of the new types with the rest of the application.

mjambon commented:
I think a good half of what's explained in your PR description should go into the source files as comments where a future visitor will find them.

kopecs committed:
Use type-directed record field disambiguation more

kopecs committed:
Add interface for Source

emjin commented on src/core/Target_location.ml:
```diff
@@ -0,0 +1,82 @@
+   * could be a tempfile
+   *
+   *)
+  file : Fpath.t;
```
I don't love the name "file"; sounds like this is closer to "contents_path"

kopecs replied:
I believe I just left it the same as the prior type used in a lot of places had so I had fewer usages to hunt down and update, but I can change it.

kopecs replied:
Done in d2a27a2ebb2cc4faec501008b313510b558cb8bd

emjin commented on src/core/Target_location.ml:
```diff
@@ -0,0 +1,82 @@
+[@@deriving show]
+
+(* A lockfile to be matched against dependency patterns. *)
+and lockfile = {
+  source : Source.t;
+  file : Fpath.t;
+  kind : Lockfile_kind.t;
+  manifest : manifest option;
+}
```
Why is this not `Lockfile_target.t`? If it's because of the lazy fields, I'd prefer `Lockfile_target` to contain two types, one (maybe `lockfile_descriptor`) with these fields, and another (`t`), with a lockfile_descriptor and the two lazy fields.

kopecs replied:
Yes, this is not `Lockfile_target.t` for the same reason `Target_location.code` is not `Xtarget.t`. See my comment [here](https://github.com/semgrep/semgrep/pull/9715#discussion_r1480677358) for more discussion of the pros/cons of defining it that way.

kopecs replied:
Resolving here to consolidate discussion on the linked thread.

emjin commented on src/core/Target_location.ml:
```diff
@@ -0,0 +1,82 @@
+(* The location of a "normal" semgrep target, comprising source code (or for
+ * regex/generic, arbitrary text data) to be executed.
+ *)
+and code = {
```
It's quite weird to me that this is called "code" but contains an optional "lockfile_target". If the intention is for us to unify the fields and turn `code` into `t`, before that happens I would rather call this something not specific to any product.

kopecs replied:
It is called "code" not really to do anything with the code product, but rather because it is source code. I want to avoid naming this `source` for obvious reasons.

kopecs committed:
Document the source module

emjin commented on src/core_scan/Core_scan.ml:
```diff
@@ -556,16 +556,18 @@ let parse_equivalences equivalences_file =
+    |> map_targets config.ncores (fun (target : Target_location.t) ->
+           let file = Target_location.file target in
+           let source = Target_location.source target in
+           logger#info "Analyzing %s (file is %s)" (Source.to_string source)
```
For these comments I would be more clear about what "file" means---maybe something like "contents written to". I want a CSE to be able to know what this means without asking us.

kopecs replied:
Done in d2a27a2ebb2cc4faec501008b313510b558cb8bd

emjin commented on src/core_scan/Core_scan.ml:
```diff
@@ -702,57 +710,60 @@ let iter_targets_and_get_matches_and_exn_to_errors (config : Core_scan_config.t)
+    manifest = manifest_target;
   }
 
+let manifest_target_location_of_input_to_code
```
Do these helpers belong here? I think I'd rather have them in Target_location or a separate module.

kopecs replied:
I believe there is a dependency issue there. This is in `Core_scan` (in `semgrep.core_scan`), and `Target_location` is in `semgrep.core`, but `Input_to_core` is available via `semgrep.targeting`. I would hesitate to move that without understanding why that is the case.

emjin commented on src/core_scan/Core_scan.mli:
```diff
@@ -86,7 +86,7 @@ val extracted_targets_of_config :
   analyzer:Xlang.t ->
   products:Semgrep_output_v1_t.product list ->
-  path:Fpath.t ->
+  source:Source.t ->
```
nit: I think it's weird that this takes a source rather than a target but I guess it's always been that way

kopecs replied:
I _could_ make it take a whole target, I suppose, but I would rather pass the minimum we really have a foreseeable need for. If it makes you happier I can rename this function.

kopecs replied:
Apparently this is exposed in the mli + used by pro so I am going to leave it the same.

emjin reviewed (commented):
I really like the explicit split between the "source" and "file". To avoid confusion, I'd like us to potentially rename "file" to be explicit about what it's for. At the very least, I'd like anything we print to the CLI to be very clear about what's happening. 

Would also like you and Matthew to coordinate to figure out the best way to organize Target_location in the context of our products. Normally I'd be open to that being in a followup PR, but that feels like the meat of this work.

kopecs committed:
Documentation for Target_location

kopecs committed:
squash! Documentation for Target_location

kopecs committed:
Use Target_location.file instead of repeating def

kopecs committed:
Call Fpath.v once in semgrep_core_with_one_pattern

kopecs committed:
Additional doc improvements (Lockfile_target)

kopecs committed:
Rename file -> internal_path_to_content

A `file` field is present for several types:
    - `Target_location.{code,lockfile,manifest}`
    - `Xtarget.t`
    - `Lockfile_target.{t,manifest}`

This renames that field to instead be `internal_path_to_content` to more
accurately reflect the usage.

kopecs committed:
fixup! Rename file -> internal_path_to_content

kopecs committed:
Use a record in Match_patterns

The various tuple elements didn't seem to be associated for the most
part, so I have put the ones which are clearly logically connected into
a record and curried the remainder.

kopecs committed:
Minor doc fixes

kopecs committed:
oops! all synopsis

mjambon commented on libs/source/Source.ml:
```diff
@@ -0,0 +1,48 @@
+(* Module for the purpose of implementing sexp conversions for t. We cannot
+ * derive sexp since Fpath does not have an sexp implementation, so we convert
+ * to a string and then derive sexp on the updated type.
+ *)
```
What are S-expressions used for? I'd rather not depend on yet another format.

kopecs replied:
I believe at some point this was getting transitively relied on somehow for the libdef serialization (which uses s-expressions). Not 100% sure through what, since I just tried to find it and it seems like it is no longer needed. Given that, I'll remove it.

kopecs replied:
Done in 5fee815255b446b1e3360b63d94021332adf6f60

mjambon commented on src/core/Target_location.mli:
```diff
@@ -0,0 +1,155 @@
+ *)
+
+(** Types for describing target locations, i.e., everything about a target
+    other than the actual contents.
```
You have to explain what a target location is. It's not a standard term that we use elsewhere.

mjambon replied:
How is it different from a source?

kopecs replied:
I think I explain what it is (perhaps without comparison or sufficient differentiation) in the comment on `type t`.

I can add more elaboration on the difference with `Source.t` though. If the question here is more than rhetorical, the distinction I have here is that a "target location" comprises sufficient information to be able to create a target (e.g., what products should rules be run from, what is the analyzer type, is there an associated lockfile, etc...) whereas a "source" is just an origin for the bytes being scanned which supply the "contents" of a target.

aryx replied:
Please again do not use the term "location" which is used for a file range/region

aryx commented on libs/source/Source.mli:
```diff
@@ -0,0 +1,60 @@
+
+(** Sources of content which may be consumed by Semgrep.
+
+   A source is the {e user-relevant location} where content (e.g., source code)
```
Please do not use location; this term is heavily charged in semgrep to mean a source code region/range.

kopecs replied:
Reworded in 5fee815255b446b1e3360b63d94021332adf6f60; that said, I do not think others necessarily have as strong an association here and I think using the word location as its plain language meaning in documentation is fine. I think it is possibly ambiguous here but I would take issue with a blanket objection to usage of location in documentation for things other than precisely byte positions within a section of source code.

In general I find it difficult to address feedback like this without at least a suggestion of the kind of alternative you would look for.

aryx commented on libs/source/Source.mli:
```diff
@@ -0,0 +1,60 @@
+
+   A source is the {e user-relevant location} where content (e.g., source code)
+   originated. This should be used when storing a location which will later be
+   reported to a user.
```
Then what is the difference between a source and a Tok.t (or Pos.t, or Loc.t).

kopecs replied:
This is "location" in a more macro sense. E.g., a specific file name; standard input; a specific network-accessible uri; a specific git object which designates a file present at a particular commit.

I will add a further note to differentiate.

kopecs replied:
Done in 194e4c0c5fc4a4ac1fab02432f90a9cd3a75d8c9.

aryx commented on libs/source/Source.mli:
```diff
@@ -0,0 +1,60 @@
+   independently of getting the content, which allows de-coupling reporting
+   where a finding originated from obtaining the contents.
+
+   This is since when these uses are fused, it is more difficult to
```
I don't understand the sentence.

kopecs replied:
Can you elaborate? It is not immediately obvious to me what to expand on based on your comment.

My best guess is that it may not be clear what "these uses" refers to. It is the actions mentioned in the preceding paragraph, namely (1) reporting where a finding originated and (2) obtaining the contents (e.g., source code) of the object from which the finding may be generated.

kopecs replied:
Reworded in 5fee815255b446b1e3360b63d94021332adf6f60.

aryx commented on libs/source/Source.mli:
```diff
@@ -0,0 +1,60 @@
+        relative path from the scanned project root.}
+    }
+ *)
+type t = File of Fpath.t [@@deriving show, eq, ord, sexp]
```
Without another variant, it's hard to understand why we need that type.
At least put in comment possible futur extensions, so we understand the need.

kopecs replied:
Added examples in 5fee815255b446b1e3360b63d94021332adf6f60

aryx commented on libs/source/Source.mli:
```diff
@@ -0,0 +1,60 @@
+ *)
+
+val to_string_opt : ?unspecified:string -> t option -> string
+(** [to_string_opt ~unspecified s] is the path [s] as a user-facing string, or
```
do not use s, it's often used for string, src seems better (if we agree that source.ml is a good name for this module).

kopecs replied:
Done in 5fee815255b446b1e3360b63d94021332adf6f60

aryx commented on src/core/Lockfile_target.ml:
```diff
@@ -1,25 +1,76 @@
-  lazy_lockfile_ast_and_errors : Dependency.t list lazy_t;
-  manifest_target : manifest_target option;
+  source : Source.t;
+      (** The source of the data as is relevant to the user. This could be, e.g., a
```
if Source.mli has good comments, and has a good name, then you would not need to repeat this comment everywhere.
But you do because Source is not a good name, and it's not clear what is its use.

aryx replied:
origin: Origin.t; seems clearer to me.

kopecs replied:
Done in 3f206d0db

aryx commented on src/core/Lockfile_target.ml:
```diff
@@ -1,25 +1,76 @@
+      associated information, or anything else a Source.t can designate.
+
+      This should be used when reporting a location to the user. *)
+  internal_path_to_content : Fpath.t;
```
the fact that both time we have this source and internal_path_to_content is a bad smell; probably there is a better abstraction to design.

kopecs replied:
Based on discussion with Martin and other comments here, I have reworked the organization of many of the additional/changed types in fc35371b4.

aryx commented on src/core/Pattern_match.ml:
```diff
@@ -90,6 +90,7 @@ type t = {
   engine_kind : Engine_kind.t; [@equal fun _a _b -> true]
   (* location info *)
   file : Fpath.t;
```
Not clear then why we need 2 fields.

kopecs replied:
Relying on changes in fc35371b4020d3fdf1beb37d3446dbe5f1e81f70 I have refactored this in 9fc64e570.

aryx commented on src/core/Xtarget.ml:
```diff
@@ -8,9 +8,36 @@
 type t = {
-  file : Fpath.t;
+  source : Source.t;
+  internal_path_to_content : Fpath.t;
```
again ... an internal_path_to_content and a source. Bad smell.

kopecs replied:
Reorg in fc35371b4 addresses this somewhat.

aryx reviewed (changes requested):
Bad names, not clear interface.
Please rename the files, source.ml is just too vague, do not use the term location (already used), and add far more comments in the .mli explaining how it differs from Tok.t, Pos.t, Loc.t, Target.t, Input_to_core.target, etc, Fpath.t.

aryx commented:
Note that for this kind of PR, I usually don't even read the modifications in the .ml. I just read the .mli and if I don't understand the comments, names, and type in this .mli, I will reject the PR.
Don't be discouraged by the many comments, back in the day brandon also did some core changes that had lots of push back. It's normal. Introducing a new name/concept/interface should be carefully reviewed.

aryx commented:
Also FYI the code in libs/ is supposed to be independent of semgrep; it's code of libraries that could be used in other projects, so try to not refer too much to semgrep in it; it's ok to use it as an example, but we can't hardcode semgrep specific stuff in it.

kopecs committed:
Address feedback for Source

kopecs committed:
squash! Address feedback for Source

kopecs committed:
fixup! Address feedback for Source

aryx commented on dune-project:
```diff
@@ -368,6 +368,11 @@ functions.
   )
 )
+(package (name source)
+  (synopsis "Types for tracking the source of scanned content")
```
I would move this code in lib_parsing/ for now, close to Loc.ml, Tok.ml, etc.
No need for one library with just one module.

kopecs replied:
Done in 3f206d0db

aryx commented on libs/source/Source.mli:
```diff
@@ -0,0 +1,77 @@
+
+(** Sources of content which may be consumed by Semgrep.
+
+   A source is the {e user-relevant origin} of content (e.g., source code).
```
I would either rename this file Src.ml, or maybe even better Origin.ml.
Note that @mjambon has actually a related type to what you did called Src_file.ml in the spacegrep codebase (libs/spacegrep/....).

kopecs replied:
Renamed in 3f206d0db; I have seen Src_file.ml but as I understand it the purpose is a bit different, since I understand that to exist for the purpose of being able to persist a copy of the contents. `Src_file.source` is potentially replaceable with `Origin.t` at some point in the future.

aryx commented on libs/source/Source.mli:
```diff
@@ -0,0 +1,77 @@
+
+   This differs from {!Pos.t} and similar types ({!Loc.t}, {!Tok.location})
+   since it does not report a location/byte position within a target, but
+   rather where {e the target} came from.
```
I think Origin.t could be a better name then, rather that source.

kopecs replied:
Renamed in 3f206d0db

aryx commented on libs/source/Source.mli:
```diff
@@ -0,0 +1,77 @@
+(** The type for sources.
+
+    {ul
+      {- [File path] sources are for files, and should have as [path] the
```
if the path must be relative, then please use Ppath.t instead of Fpath.t (in libs/paths/)
so we can enforce it.

aryx commented on libs/source/Source.mli:
```diff
@@ -0,0 +1,77 @@
+    Possible future variants:
+
+    {ul
+      {- [Stdin] sources are for content read from standard input.}
```
@mjambon actually had such a variant in his Src_file.t type.

aryx commented on src/core/Lockfile_target.ml:
```diff
@@ -1,25 +1,76 @@
+      associated information, or anything else a Source.t can designate.
+
+      This should be used when reporting a location to the user. *)
+  internal_path_to_content : Fpath.t;
```
I still see those source: xxx; internal_path_to_content: xxx everywhere, which is bad smell.

kopecs replied:
Addressed in fc35371b4

aryx commented:
The added comments were useful. Thanks, but I think we still need improvements.

kopecs committed:
Reorganise target types

Addressing PR feedback from Pad and Martin, this reorganises changes:

    - Since `Target_location` is a poor name, and this PR already
      removed the vestigial `Target` module (just a type alias for
      AST_generic.t), the types from `Target_location` are moved to
      `Target`.

      This creates better alignment with `Xtarget` and
      `Input_to_core.target`.
    - Similarly, since `Target` is now used consistently for
      "information about files we are going to scan, notwithstanding the
      contents thereof", `Lockfile_target` is renamed
      `Lockfile_xtarget`, since it is already parallel to `Xtarget`.
    - There is also some restructing of the types in (now) `Target`,
      `Xtarget` and `Lockfile_xtarget`. Some shared fields are now
      factored out into common types (see, e.g., `Target.target_path`),
      although this is not possible in all cases. As an example, see the
      repetition of `analyzer`/`xlang` in `Target.code`/`Xtarget.t`.
      `Xtarget.t` does not have a field which is a `Target.code` since
      this would require it to carry around a `Target.lockfile` as well.
      If we determine it is better to re-use the type at the cost of
      carrying around this additional information, we can revise this.

kopecs committed:
Rename Source to Origin

This also moves it to lib_parsing at Pad's suggestion.

kopecs committed:
Use `Target.target_path` where appropriate

Now that we have this type broken out, we can use it in more places,
notably `Pattern_match.t` and to replace the `target_info` type used in
`Xpattern_matcher`.

kopecs committed:
squash! Reorganise target types

aryx commented on js/engine/Main.ml:
```diff
@@ -48,15 +48,9 @@ let _ =
-                     })
+               (fun f : Target.t ->
+                 Code
+                   (Target.code_of_origin xlang Product.all (File (Fpath.v f))))
```
code_of_origin is a bad name given you don't pass just an origin here but lots of arguments.

kopecs replied:
Done in 1e38a3c65ba5f117024690f171ed62c91a1ad09f. I have called this functions`mk_code` instead (and likewise for the other versions). Please let me know a more concrete suggestion if that's also disliked.

aryx commented on libs/lib_parsing/Origin.mli:
```diff
@@ -0,0 +1,77 @@
+
+(** Origins of content which may be consumed by Semgrep.
+
+   A source is the {e user-relevant origin} of content (e.g., source code).
```
Need to update all the comment to not use "source" anymore but "origin".

kopecs replied:
Thought I got them all 😔

kopecs replied:
Done in 288cddf0d3fc44b07b1539f997a425128dceaeb6

aryx commented on src/core/Target.ml:
```diff
@@ -12,18 +12,52 @@
-(*****************************************************************************)
+(* See Target.mli for documentation of public items. *)
+
+type target_path = { origin : Origin.t; internal_path_to_content : Fpath.t }
```
since the file is already called Target.ml, you can rename target_path to just path, so qualified it looks like Target.path.

aryx replied:
Also when the origin will not be a File, what will be internal_path_to_content? A path to a local tmp file?
Then would be good to add such a comment, explaining the difference
between the File path.t in origin and internal_path_to_content.

kopecs replied:
I believe this is addressed in the documentation in the mli: https://github.com/semgrep/semgrep/blob/176e9e6b6aa263d12c9dc52180293c9505ba25c9/src/core/Target.mli#L76-L95

are you looking for something beyond what is said there?

kopecs replied:
First comment done in b8ee62ef273f8e3520d1865224989e6ccac03d66

aryx commented on src/core/Target.mli:
```diff
@@ -0,0 +1,144 @@
+    However, it does not contain the actual contents (parsed or otherwise) of
+    the target itself.
+ *)
+type t = Code of code | Lockfile of lockfile [@@deriving show]
```
Code is a very bad name, as this data structure does not contain code.
Maybe call it RegularTarget,

kopecs replied:
This is called `Code` so that it is consistent with the types @mmcqd introduced in https://github.com/semgrep/semgrep/pull/9695. I can change it here, but it would then no longer be consistent with the similar type added there.

I can rename both, and update the python code which generates this, but I am uncertain that this PR is the appropriate place to do so.

mmcqd replied:
I'm happy to change the name. Just wanted to indicate it's a target with source code in it and we should do regular matching on it. `RegularTarget` seems good to me. If Cooper is ok with changing the name here that's fine with me, or I can I open a separate PR after this one renaming everything.

kopecs replied:
I have renamed this `Regular`, without the Target, (given it is in the module name already) in e18e0a6b4e5b7776a80ac46ff158b1490806f156.

aryx commented on src/core/Target.mli:
```diff
@@ -0,0 +1,144 @@
+    contents of [target]. *)
+
+val origin : t -> Origin.t
+(** [origin target] is the user-reportable origin of [target]. *)
```
what are those empty Target_location.ml files below?

kopecs replied:
These were the moved files which became Target.ml; for some reason they must still be in the index as empty. I will remove them.

kopecs replied:
Cleaned up in 8c905bd51f56312325aaa7761cfc80868b953845

kopecs committed:
Ensure Target_location is removed

kopecs committed:
Rename functions from x_of_origin -> mk_x

These take sufficient additional parameters, per Pad, to deserve a
different name.

kopecs committed:
squash! Rename Source to Origin

kopecs committed:
Rename `Target.target_path` -> `Target.path`

kopecs committed:
Fix broken doc links

aryx commented on libs/lib_parsing/Origin.mli:
```diff
@@ -0,0 +1,77 @@
+
+ *)
+
+(** The type for origns.
```
typo

kopecs replied:
Fixed in 3e5eb80172398b55b3f77f4747befea521059095.

aryx commented on src/core/Target.mli:
```diff
@@ -0,0 +1,146 @@
+    pysemgrep generates targets that have slightly less information (e.g.,
+    these types have expanded information about the targets' locations). *)
+
+(** The location of a Semgrep target. This contains all of the target-specific
```
Seems like it's actually not just the location of a Semgrep target, so I would remove this leading comment.

kopecs replied:
Ah yes, this is just an artifact of the fact that this was previously `Target_location`. I'll revise to make sure this is consistent with the Target/Xtarget naming.

kopecs replied:
Fixed in f8c7725030537be263c25510c3a0cc5f92a5d510.

aryx commented on src/core/Target.mli:
```diff
@@ -0,0 +1,146 @@
+(** A manifest file to be used during matching. See also
+    {!Lockfile_xtarget.manifest}, which also has the contents. *)
+
+and path = {
```
please avoid mutually recursive construct like 'and' when the type is actually not mutually recursive.
I know sometimes it's nice to put the important type first, and abuse 'and' so that one can order
the types as we please, but better to actually see when types really need to be mutually recursive.

kopecs replied:
I have made this change in 275903281, but I think it is a net-negative. We already have other definitions (e.g., [various taint types for Rule.t](https://github.com/semgrep/semgrep/blob/55412a8d87addd85ec1543c82a4598f4b44847df/src/core/Rule.ml#L168-L304)) defined with `and` which are not mutually recursive, and I think listing the types in what is essentially the reverse order of relevancy simply makes the interface harder to read (further hurt by the fact that you then end up with the most important types sandwiched between other types and functions, not even at the end of the file).

aryx commented on js/engine/Main.ml:
```diff
@@ -48,15 +48,8 @@ let _ =
-                       lockfile_target = None;
-                     })
+               (fun f : Target.t ->
+                 Code (Target.mk_code xlang Product.all (File (Fpath.v f))))
```
I still think 'code' is bad. I'd rather have mk_regular_target and mk_lockfile_target.
cc @mmcqd for ideas for better names for this.

mmcqd replied:
"regular target" and "lockfile target" sound good to me

kopecs replied:
I renamed this to `mk_regular` as part of the change in e18e0a6b4e5b7776a80ac46ff158b1490806f156.

Given these are always used like `Target.mk_regular` or `Target.mk_lockfile` I don't see the point of including `target` again in the function name (similar to `Target.path` being preferred to `Target.target_path`).

aryx commented on src/core/Target.mli:
```diff
@@ -0,0 +1,146 @@
+    {- obtaining the contents: [internal_path_to_content]}
+  } *)
+
+(* deriving eq appears to not work on mutual definitions with and. See also
```
weird, Iwe have lots of recursive types in AST_generic.ml and we derive show, eq, fine.

kopecs replied:
Show appears to work fine, and I think eq is fine when applied to _all_ types, but in the case where you have
```
type t [@@deriving show, eq]
and q [@@deriving show]
```
it appears that `equal : t -> t -> bool` is not generated.

aryx replied:
because for recursive types you just need the deriving at the very end, after all the and.

aryx commented on src/core/Target.mli:
```diff
@@ -0,0 +1,146 @@
+  Origin.t ->
+  code
+(** [mk_code analyzer products origin] is the target
+      location for a source code target originating from [origin] to be
```
then maybe call it mk_source_code_target

kopecs replied:
I renamed this in e18e0a6b4e5b7776a80ac46ff158b1490806f156 to `mk_regular`. In general I think this should parallel the name of the constructor for `Target.t`. If you disagree please let me know, but I hesitate to break that correspondence without a compelling reason.

aryx commented on src/core/Target.mli:
```diff
@@ -0,0 +1,146 @@
+      a target from certain types of origins, such as generating a tempfile.
+ *)
+
+val mk_lockfile : ?manifest:manifest -> Lockfile_kind.t -> Origin.t -> lockfile
```
and mk_lockfile_target (or sca_target).

kopecs replied:
I think this should still be `mk_lockfile` for a couple reasons:
- I think this should parallel the name of the constructor for Target.t. Currently that is `Lockfile`, carried over from Matthew's recent PR.
- Given these are always used like Target.mk_regular or Target.mk_lockfile I don't see the point of including target again in the function name (similar to Target.path being preferred to Target.target_path).

aryx commented on src/core/Xtarget.ml:
```diff
@@ -1,16 +1,56 @@
 type t = {
-  file : Fpath.t;
-  xlang : Xlang.t;
+  path : Target.path;
```
at this point maybe we could wrap completely Target.t instead of Target.path,
since Target.t contains also xlang that we need for Xtarget.

kopecs replied:
The unfortunate issue there is that you then must carry around the `Target.lockfile`, if one is associated. Ideally we would carry the `Lockfile_xtarget.t` around instead (but in any case, we do not currently need to access it from the `Xtarget` and so such a field does not exist here).

I can refactor it to do that, but I anticipate it would simply add to memory consumption for rather marginal benefit.

aryx reviewed (changes requested):
Need to rebase and solve the conflicts, but this looks better.
I think we're close to merge this.

aryx commented:
I think I can probably piggiback on your Target.t at some point and add capabilities to them,
so once you have a Target.t, you'll have the capability to read the file (even though the content is not there yet, it will
be in Xtarget.ml), and most of the engine will not have to use open and will remain pure (capability wise).

kopecs committed:
Target documentation cleanup

Remove outstanding artefactual references to "locations".

kopecs committed:
fixup! squash! Rename Source to Origin

kopecs committed:
Rename `Target.Code` to `Target.Regular`

kopecs committed:
Do not use `and` in Target.mli

kopecs committed:
Merge with develop; mostly Logs related conflicts

Martin's recent replacement of Easy_logging with Logs caused a couple
conflicts.

aryx commented:
CI failures

aryx commented on js/engine/Main.ml:
```diff
@@ -48,15 +48,9 @@ let _ =
-                     })
+               (fun f : Target.t ->
+                 Regular
+                   (Target.mk_regular xlang Product.all (File (Fpath.v f))))
```
I guess mk_regular could also handle the Regular constructor, so this
whole thing become simply a Target.mk_regular ... (without the redundant Regular constructor around).

kopecs replied:
The only qualm I have with that is that then trying to make a `Target.regular` is more painful. E.g, a usage like [this](https://github.com/semgrep/semgrep/blob/2852c5771fd5a447601064aa1757623d42d92bf7/src/core_scan/Core_scan.ml#L750) becomes very unwieldy.

aryx commented on libs/lib_parsing/Origin.mli:
```diff
@@ -0,0 +1,77 @@
+
+    {ul
+      {- [File path] origins are for files, and should have as [path] the
+        relative path from the scanned project root.}
```
add a 'TODO: use Ppath.t at some point'

kopecs replied:
Done in d57830ad536d2800a7e4c53159e6c156191eae83.

aryx commented on src/core/Lockfile_xtarget.ml:
```diff
@@ -0,0 +1,67 @@
+  manifest : manifest option;
+  lazy_content : string lazy_t;
+      (** The contents of the lockfile, as a string. *)
+  lazy_ast_and_errors : Dependency.t list lazy_t;
```
Maybe rename this field since this does not really contain ast and errors right?

kopecs replied:
Done in d57830ad536d2800a7e4c53159e6c156191eae83.

aryx commented on src/core/Lockfile_xtarget.ml:
```diff
@@ -0,0 +1,67 @@
+  target : Target.manifest;
+  lazy_content : string lazy_t;
+      (** The contents of the manifest, as a string. *)
+  lazy_ast_and_errors : Dependency.manifest_dependency list lazy_t;
```
same, I would rename the field, or if it's because you want to mimic Xtarget.ml,
then maybe we need to make this type polymorphic so we can actually factorize things.

kopecs replied:
>  I would rename the field

Done in d57830ad536d2800a7e4c53159e6c156191eae83.

> then maybe we need to make this type polymorphic so we can actually factorize things.

I considered this but @mjambon was against using polymorphism for factorization of these types.

aryx commented on src/core/Target.ml:
```diff
@@ -12,18 +12,52 @@
+
+let path_of_origin (origin : Origin.t) : path =
+  match origin with
+  | File file -> { origin; internal_path_to_content = file }
```
add a TODO saying at some point here we will create temporary files for complex origin like GitBlob

kopecs replied:
Done in d57830ad536d2800a7e4c53159e6c156191eae83.

aryx commented on src/core/Target.ml:
```diff
@@ -12,18 +12,52 @@
+}
+[@@deriving show]
+
+type t = Regular of regular | Lockfile of lockfile [@@deriving show]
```
In theory because both variants have this path field, you could move it up to factorize.
But maybe it's fine like this.

kopecs replied:
There are some other threads with similar suggestions, but I do not think that works since we need to be able to distinguish later between regular and lockfile targets (e.g., [here](https://github.com/semgrep/semgrep/pull/9715/files#diff-b3fa3bcb465c63ed55d1e0a69a161dbff7cffc688b30a1750caf9f870ed8d555R69-R73)). If this were instead
```ocaml
type u = Regular of regular | Lockfile of lockfile (* without path field for either *)
type t = { path : path; target : u }
```
then if I am passing around a value of `Target.regular` I also need to separately keep the path, and if I use a value of type `t` then I cannot enforce at the type level it is particularly a regular or lockfile target.

Something like
```ocaml
type 'a t = { path : path; target : 'a }
type regular = regular_info t;
type lockfile = lockfile_info t;
type any = (regular_info, lockfile_info) Either.t t;
```
could work, but Martin was of the opinion such a pattern should be avoided due to error messages & inference.

aryx commented on src/core/Xtarget.ml:
```diff
@@ -1,16 +1,56 @@
+      (** This is valid only for xlang = Xlang.L ..., not for LRegex|LGeneric *)
 }
+
+let parse_file parser (analyzer : Xlang.t) path =
```
now that this type has a few functions, please add a Xtarget.mli

kopecs replied:
Done in d57830ad536d2800a7e4c53159e6c156191eae83.

aryx commented on src/core/Target.mli:
```diff
@@ -0,0 +1,140 @@
+    target from certain types of origins, such as generating a tempfile.
+ *)
+
+val internal_path_to_content : t -> Fpath.t
```
rename just path (or fpath), otherwise call sites are harder to read.
(you can keep the internal_path_to_content field name though)

kopecs replied:
How's `internal_path`? I would prefer something which at least indicates this may not be meaningful to the user, since I could see people potentially using `path` without thinking about it.

aryx replied:
yes, I like internal_path

kopecs replied:
Done in d57830ad536d2800a7e4c53159e6c156191eae83.

aryx commented on src/core_cli/Core_command.ml:
```diff
@@ -189,9 +189,8 @@ let semgrep_core_with_one_pattern (config : Core_scan_config.t) : unit =
-             | `CodeTarget (t : Input_to_core_t.code_target) -> t.path
-             | `LockfileTarget (t : Input_to_core_t.lockfile_target) -> t.path)
+        |> List_.map (fun (t : Target.t) ->
+               Target.internal_path_to_content t |> Fpath.to_string)
```
would be easier to read as Target.path t |> Fpath.to_string.

Also we should probably use Fpath.t for files, so we don't beed Fpath.to_string.

kopecs replied:
> would be easier to read as Target.path t |> Fpath.to_string.

Addressing this at https://github.com/semgrep/semgrep/pull/9715/files/2852c5771fd5a447601064aa1757623d42d92bf7#r1489106080

> Also we should probably use Fpath.t for files, so we don't beed Fpath.to_string.

See my earlier comment here: https://github.com/semgrep/semgrep/pull/9715#discussion_r1481932234.

aryx commented on src/core_cli/Core_command.ml:
```diff
@@ -206,14 +205,15 @@ let semgrep_core_with_one_pattern (config : Core_scan_config.t) : unit =
                      Parse_target.parse_and_resolve_name_warn_if_partial lang
                        file
                    in
+                   let path = Fpath.v file in
```
see my comment above, if we use directly Fpath.t for files, we don't need to convert back again to Fpath.t here.

You can use !! also instead of Fpath.to_string if the path is used in format string %s

kopecs replied:
Discussion on potential issues with replacing `Fpath.t` with `string` here at https://github.com/semgrep/semgrep/pull/9715/files/2852c5771fd5a447601064aa1757623d42d92bf7#r1489954087

aryx commented on libs/commons/List_.ml:
```diff
@@ -325,6 +325,11 @@ let optlist_to_list = function
 
 let sort xs = List.sort compare xs
 
+let sort_by_key key cmp xs =
```
maybe I would rename sort_by_mapped_key,
to avoid ambiguity with the Assoc_.sort_by_key_xxx function
and I would rename key as fkey, so it's clear we're passing a function here, not a key.

kopecs replied:
>  I would rename key as fkey, so it's clear we're passing a function here, not a key.

I am not certain what would be unclear here, given the function is 3 lines and the one usage of this is that it is applied. I can rename it, but I think this probably speaks more to the fact that this warrants documentation. I will add some in the mli.

> maybe I would rename sort_by_mapped_key,
to avoid ambiguity with the Assoc_.sort_by_key_xxx function

I would rather not rename this, since I think the suggestion simply increases verbosity and imo any ambiguity is resolved by the module name.

aryx reviewed (changes requested):
Looking good.

IagoAbal reviewed (approved):
Just approving the changes to taint code, will review the rest too and add comments if needed.

kopecs committed:
Address feedback

kopecs committed:
Fix factorization of xtarget_of_file for tests

github-actions[bot] committed:
Update pytest snapshots

kopecs commented:
CI failure should now just be pro. There is a corresponding PR [here](https://github.com/semgrep/semgrep-proprietary/pull/1291).

aryx commented:
@kopecs maybe we could start to use Origin.t and extend it slightly to handle parse errors in patterns inside rules.
Right now we have those many parse_pattern in languages/, and actually ocaml-tree-sitter-core accept a ?src_file argument
to pass the origin of content, so maybe we could use that so that errors in patterns inside rules are better reported.

kopecs commented:
Yes, I would like to. I can look into it a bit but I am a bit bogged down with other things this week.
